Add metrics for scheduler framework.

liu-cong opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind flake

/kind feature

**What this PR does / why we need it**:
#83416 

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #83416

**Special notes for your reviewer**:
In terms of overhead of `time.Now()` for each plugin, I ran a simple go program which calls `time.Now()` many times and then log the duration, the result is

MacOs, 3.1 GHz Intel Core i7, 16GB RAM

\# calls | latency
-- |-- 
10 | 1.15 µs
100 | 10.58 µs
1000 | 102.666µs

Suppose we have 100 plugins to run, it adds ~10us to each scheduling cycle. Given benchmark results with tens or hundreds **ms/op**, I think the overhead is negligible

**Does this PR introduce a user-facing change?**:
New metrics will be added.
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add latency and request count metrics for scheduler framework.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig scheduling

liu-cong committed:
Add metrics for scheduler framework.

liu-cong commented:
/test pull-kubernetes-kubemark-e2e-gce-big

ahg-g commented:
Can you please add an option to NewFramework to enable/disable tracing? Something like WithTracing(bool), the default should be disabled for now until we have few plugins implemented and allow us to do benchmarking of the tracing overhead on real workloads.

liu-cong commented:
> Can you please add an option to NewFramework to enable/disable tracing? Something like WithTracing(bool), the default should be disabled for now until we have few plugins implemented and allow us to do benchmarking of the tracing overhead on real workloads.

Are you suggesting adding this option as a temporal approach to make benchmarking easier as we add more plugins? Or to add a permanent flag?

I have relatively high confidence that the overhead is negligible. However if you feel not confident without benchmarking with real plugins, I think we can skip the per-plugin tracing metrics for now and add it back once we converted all plugins. I prefer not to introduce an option if we (very likely) don't need it.

Huang-Wei commented:
@liu-cong We may need sig-instrumentation to help do another review here.

I will do a first round of review.

Huang-Wei reviewed (commented):
Thanks @liu-cong , some nits, LGTM otherwise.

liu-cong commented:
> > Can you please add an option to NewFramework to enable/disable tracing? Something like WithTracing(bool), the default should be disabled for now until we have few plugins implemented and allow us to do benchmarking of the tracing overhead on real workloads.
> 
> Are you suggesting adding this option as a temporal approach to make benchmarking easier as we add more plugins? Or to add a permanent flag?
> 
> I have relatively high confidence that the overhead is negligible. However if you feel not confident without benchmarking with real plugins, I think we can skip the per-plugin tracing metrics for now and add it back once we converted all plugins. I prefer not to introduce an option if we (very likely) don't need it.

Removed per-plugin metrics for now.

> > > Can you please add an option to NewFramework to enable/disable tracing? Something like WithTracing(bool), the default should be disabled for now until we have few plugins implemented and allow us to do benchmarking of the tracing overhead on real workloads.
> > 
> > 
> > Are you suggesting adding this option as a temporal approach to make benchmarking easier as we add more plugins? Or to add a permanent flag?
> > I have relatively high confidence that the overhead is negligible. However if you feel not confident without benchmarking with real plugins, I think we can skip the per-plugin tracing metrics for now and add it back once we converted all plugins. I prefer not to introduce an option if we (very likely) don't need it.
> 
> as we discussed offline, lets only keep the per extension point tracing and leave the per-plugin one until we have enough plugins that will allow us to benchmark the impact of per-plugin tracing.

Removed per-plugin metrics for now.

liu-cong commented:
/test pull-kubernetes-integration

Huang-Wei commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -222,7 +235,9 @@ func (f *framework) QueueSortFunc() LessFunc {
-	state *CycleState, pod *v1.Pod) *Status {
+	state *CycleState, pod *v1.Pod) (status *Status) {
+	startTime := time.Now()
+	defer recordExtensionPointDuration(startTime, preFilter, &status)
```
Have you tested it? I think `recordExtensionPointDuration` will always get `status` as nil, isn't it?

Huang-Wei replied:
I don't think using named return variable is a good practice here.

liu-cong replied:
yes please take a look at the unit tests.
The purpose of having a named return variable is to make the `defer` statement easy to execute. Otherwise we have to make sure the `recordExtensionPointDurationc` called properly in various control branches, which is more error prone.

Huang-Wei commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -604,3 +641,13 @@ func (f *framework) pluginsNeeded(plugins *config.Plugins) map[string]config.Plu
+
+// status is a double pointer to Status because Status is not created when arguments
+// of this function is evaluated in "defer".
+func recordExtensionPointDuration(start time.Time, extensionPoint string, status **Status) {
```
Why doulbe pointer is necessary?

liu-cong replied:
because the arguments to deferred functions are evaluated immediately. Without the double pointer, it will always be nil.

liu-cong replied:
Refactored the code to avoid using double pointer.

ahg-g commented:
/lgtm

I will leave it to @Huang-Wei  to approve.

Huang-Wei commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -222,7 +235,9 @@ func (f *framework) QueueSortFunc() LessFunc {
-	state *CycleState, pod *v1.Pod) *Status {
+	state *CycleState, pod *v1.Pod) (status *Status) {
+	startTime := time.Now()
+	defer func() { recordExtensionPointDuration(startTime, preFilter, status) }()
```
Putting it into a closure so that we can always get the correct `status` instead of zero value. 👍

Huang-Wei commented:
/approve

Thanks, @liu-cong !
